src/Components/Endpoints/src/Binding/DefaultFormValuesSupplier.cs

```diff
@@ -14,69 +14,96 @@ namespace Microsoft.AspNetCore.Components.Endpoints;
-            boundValue = result;
-            return true;
-        }
-
-        boundValue = valueType.IsValueType ? Activator.CreateInstance(valueType) : null;
-        return false;
+        deserializer.Deserialize(context, _options, _formData.Entries);
     }
 
-    private Func<IReadOnlyDictionary<string, StringValues>, FormDataMapperOptions, string, object> CreateDeserializer(Type type) =>
-        _method.MakeGenericMethod(type)
-        .CreateDelegate<Func<IReadOnlyDictionary<string, StringValues>, FormDataMapperOptions, string, object>>();
+    private FormValueSupplier CreateDeserializer(Type type) =>
+        (FormValueSupplier)Activator.CreateInstance(typeof(FormValueSupplier<>)
+        .MakeGenericType(type))!;
 
-    private static object? DeserializeCore<T>(IReadOnlyDictionary<string, StringValues> form, FormDataMapperOptions options, string value)
+    internal abstract class FormValueSupplier
     {
-        // Form values are parsed according to the culture of the request, which is set to the current culture by the localization middleware.
-        // Some form input types use the invariant culture when sending the data to the server. For those cases, we'll
-        // provide a way to override the culture to use to parse that value.
-        var reader = new FormDataReader(form, CultureInfo.CurrentCulture);
-        reader.PushPrefix(value);
-        return FormDataMapper.Map<T>(reader, options);
+        public abstract void Deserialize(FormValueSupplierContext context, FormDataMapperOptions options, IReadOnlyDictionary<string, StringValues> form);
     }
 
-    public bool CanConvertSingleValue(Type type)
+    internal class FormValueSupplier<T> : FormValueSupplier
     {
-        return _options.IsSingleValueConverter(type);
+        public override void Deserialize(FormValueSupplierContext context, FormDataMapperOptions options, IReadOnlyDictionary<string, StringValues> form)
+        {
+            if (form.Count == 0)
+            {
+                return;
+            }
+
+            char[]? buffer = null;
+            try
+            {
+                var maxKeyLenght = -1;
+                var dictionary = new Dictionary<FormKey, StringValues>();
+                foreach (var (key, value) in form)
+                {
+                    if (key.Length > maxKeyLenght)
+                    {
+                        maxKeyLenght = key.Length;
+                    }
+                    dictionary.Add(new FormKey(key.AsMemory()), value);
+                }
+                buffer = ArrayPool<char>.Shared.Rent(maxKeyLenght);
+
+                // Form values are parsed according to the culture of the request, which is set to the current culture by the localization middleware.
+                // Some form input types use the invariant culture when sending the data to the server. For those cases, we'll
+                // provide a way to override the culture to use to parse that value.
```

Comment:
Interesting - does this mean some public API for it? Where will it go and what will it look like?

Reply:
This will be part of the binder API, as an attribute, that you apply on a specific object property.

MVC does a thing where it emits a hidden field to signal the input needs to be parsed in a very specific way, but I am not a fan of that.

Reply:
I see, thanks.

> MVC does a thing where it emits a hidden field to signal the input needs to be parsed in a very specific way, but I am not a fan of that.

It makes sense why MVC's approach could be regarded as the correct solution, since a single model field may be used in multiple different UI scenarios, and only some of those UI scenarios might involve culture-specific parsing. However it's really messy for the developer if their HTML forms get filled up with system-generated boilerplate to describe things like parsing rules. I can see why you might prefer to tell people it's cleaner to create a form-specific DTO so it can be annotated.

Overall my sense is that no solution is going to delight everyone in all cases, but sensible defaults might cover a lot of the scenarios. For example, assuming culture-invariant parsing for any time that obviously maps to a standard HTML input type that produces culture-invariant values (like `DateTime`), since the most mainstream use case would be to use the corresponding HTML input type.

Anyway that's all outside the scope of this PR.

Reply:
> Overall my sense is that no solution is going to delight everyone in all cases, but sensible defaults might cover a lot of the scenarios. For example, assuming culture-invariant parsing for any type that obviously maps to a standard HTML input type that produces culture-invariant values (like DateTime), since the most mainstream use case would be to use the corresponding HTML input type.

The challenge here is that you don't know what input that value is bound to while you are doing the mapping. It might be an HTML5 input, or it can be a totally custom control provided by a third-party library. This impacts non-english cultures the most, for example, in spanish the decimal separator is the `,` as opposed to the `.` in English.

Reply:
> This impacts non-english cultures the most, for example, in spanish the decimal separator is the , as opposed to the . in English.

I know, that's what I was saying too. We would have to pick a default, and one possible default is the assumption that - in your example - a numeric value would have come from an `<input type=number>` and hence arrive in a culture-invariant format. Of course we don't know for sure if that is the type of UI that was used in any individual case, but we have to pick a default one way or the other.

Reply:
I picked a culture sensitive default since that "aligns" with what MVC does today and will ease up the migrations. There are other alternative approaches we can pursue, since we run JS on the page, we can always choose to process the value and convert it into the given culture using the underlying JS INTL APIs before we submit the form.

We'll also make the option of which culture to use something you can configure, so you can choose to perform invariant conversions all the time.

Reply:
Can you make a more convincing case that this is the right thing to do? It doesn't seem right to me because Blazor already has an established convention around binding, specified in the various `Input*` components and `BindConverter.cs`. For example, `InputNumber` works regardless of the user's browser culture, since `<input type=number>` supplies culture-invariant values, and `InputNumber` does culture-invariant parsing. I know the server doesn't know what HTML element type the user used, so it needs to pick a default.

The culture-sensitive parsing mechanism proposed here looks like it would not align with the most basic conventional form setup for Blazor and would give incorrect behavior by default in something like https://blazorrepl.telerik.com/cnuAQglu197hQ2cY09. For example, if there's an `<input type=number step=0.001>`, then if the server culture is Spanish, and a Spanish user types `123,456`, the posted value will be `123.456` which will either be an error (if binding to an `int`) or will be recorded dangerously as a wrong value (if binding to a `double`). If the app was developed by an English-language team, it's also quite possible they would have no idea that this correctness issue could exist as it always does what they expect in their language, so they would not know to change any defaults.

I understand there's prior art in MVC, and there's different prior art in Blazor. I'm not trying to argue for the sake of it here; I'm trying to get to a point of confidence that we've made the right choice for the defaults (even when there are flags and such that can be used to override it).

Reply:
I don't have strong feelings for either/or. I picked the current culture because that's what MVC did already, but I am fine if we change it to be invariant culture by default. In the end, this will need to be configurable.

Reply:
@SteveSandersonMS what do you prefer.

Reply:
Thanks for being open to this!

Looking at the rules in `BindAttributes.cs`, there is one case where we definitely won't be able to recreate how interactive binding works, whichever default we pick. That is, Blazor distinguishes between `<input type=text>` (culture sensitive) and `<input type=number/date/datetime-local/month/time>` (culture insensitive). The server can't know which of those was used, so no matter which way the default goes, it will be possible to construct a case where the behavior of SSR form binding will differ from interactive binding.

So I guess we have to think of which kind of behavior is more commonly wanted, and how bad are the consequences of getting it wrong:

 * [A] If we go insensitive by default, all the specialized types (`number/date/datetime-local/month/time`) will work out of the box regardless of user browser culture. But if someone uses `<input type=text>`:
   * [A1] If the server culture does match the user culture, the developer might expect some culture-sensitive parsing to occur but it won't
   * [A2] If the server culture doesn't match the user culture, then this was never going to produce a useful result, and invariant parsing is probably safer and more predictable than parsing with a wrong culture
 * [B] If we go sensitive by default, all the specialized types will only work on sufficiently English-like cultures since it has the privilege of being the same as "invariant", whereas in other cultures `number/date/datetime-local/month/time` will be broken. And as for `<input type=text>`:
   * [B1] If the server culture does match the user culture, then culture-specific parsing will occur as desired
   * [B2] If the server culture doesn't match the user culture, then like above, unpredictable behavior ensues

So on balance I'd see:

 * insensitive default is better for `number/date/datetime-local/month/time`
 * sensitive default is better for `<input type=text>` binding to a nonstring value, but only if the server culture matches the user culture
 * neither is better for `<input type=text>` binding to a nonstring value if the server culture doesn't match the user culture

... and with that I'd lead towards insensitive default as handling a wider range of scenarios.

That's my initial reasoning anyway but am happy to be corrected if I'm missing key scenarios!

Reply:
I've updated this to use culture invariant by default, there will be an option to switch to current culture instead.